### PR TITLE
Reduce allocations when producing checksums.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Saves about 100MB of allocations when on NetCore6 or above for a simple typing/lightbulb scenario: 

image

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs [hunk 1]
```diff
@@ -98,33 +105,90 @@ public static Checksum Create(IObjectWritable @object)
...
+#if NET
+            return CreateUsingSpans(checksum1, checksum2, checksum3);
+#else
+            return CrateUsingByteArrays(checksum1, checksum2, checksum3);
+#endif
+        }
 
-            stream.Position = 0;
-            return Create(stream);
+        private static Checksum CrateUsingByteArrays(Checksum checksum1, Checksum checksum2)
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:124]: >CrateUsingByteArrays [](http://example.com/codeflow?start=32&length=20)

nit: typo

Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:124]: Should the byteArray methods be in a #else for a #if NET?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs]: Fixed.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs]: if i do that, i can't then write a test that validates them against each other... (Definitely looking for clever ideas here.  Just not sure what they are!)

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs [hunk 2]
```diff
@@ -98,33 +105,90 @@ public static Checksum Create(IObjectWritable @object)
...
+#if NET
+
+        // Optimized helpers that do not need to allocate any arrays to combine hashes.
+
+        private static Checksum CreateUsingSpans(Checksum checksum1, Checksum checksum2)
+        {
+            using var hash = s_incrementalHashPool.GetPooledObject();
+
+            Span<byte> bytesSpan = stackalloc byte[2 * HashSize];
+            Span<byte> hashResultSpan = stackalloc byte[hash.Object.HashLengthInBytes];
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:162]: >Span<byte> hashResultSpan = stackalloc byte[hash.Object.HashLengthInBytes]; [](http://example.com/codeflow?start=12&length=75)

Is it worth the complexity to have the separate implementation for these? I feel like the pooled object one will have very similar performance numbers amortized.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:162]: I'm amenable to either approach.  But this all seems self contained and nice to not even hit the heap.  Basically, we have a few 20 byte numbers and we're producing a final 20 byte number. Why not just have that work like normal math would work with ints on the stack:-)

Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:162]: I'm missing something obvious, what's the reason the bytearray methods can't use stackalloc? I get that they can't use the new GetHashAndReset, I'm just wondering why they differ in the byte array allocation code.

Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:162]: thanks for the explanation, haven't used stackalloc much and that helps

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**